server.py

```diff
@@ -61,28 +67,78 @@ def parse_request(request):
         return uri
 
 
-def setup_server():
-    server_socket = socket.socket(
-        socket.AF_INET, socket.SOCK_STREAM, socket.IPPROTO_IP)
-    server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-    server_socket.bind(ADDR)
-    server_socket.listen(1)
-    return server_socket
-
-
 def response_ok(uri):
     """Return a status 200 HTTP response_ok"""
-
+    message, ctype = resolve_uri(uri)
     return RESPONSE.format(response_code=b'200',
-                           response_reason=b'OK', date=foo_date)
+                           response_reason=b'OK', date=foo_date,
+                           ctype=ctype, message=message)
 
 
 def response_error(code, reason_phrase):
-    """Return a status 500 Internal Server Error"""
+    """Return an error response"""
+
+    return RESPONSE.format(response_code=code,
+                           response_reason=reason_phrase,
+                           date=foo_date, ctype=b'text/plain', message="")
 
-    return RESPONSE.format(response_code=b'500',
-                           response_reason=b'OK', date=foo_date)
 
+def resolve_uri(uri):
+    """Takes a uri and returns the contents a tuple of (body, content-type)
+
+    body:
+    if uri points to a file, returns the contents of that file,
+    if uri points to a dir, returns an html file outlining contents
+
+    content-type:
+    when possible, will return a content type that is consistent with the
+    content-type HTTP header.
+
+    invalid requests will raise an appropriate Python exception
+
+    """
+    if len(uri.split(b'//', 1)) == 2:
+        # Case of absolute uri
+        pth_lst = uri.split(b'/')[3:]  # Throw out 'http://www.anyhost.com' bits
+    else:
+        # Case of relative uri
+        pth_lst = uri.split(b'/')
+        if pth_lst[0] in set([b"", b"."]):
+            # Case of starting path with either "/" or "./"
+            pth_lst = pth_lst[1:]
+    pth = os.path.join(ROOT, *pth_lst)
+
+    if os.path.isdir(pth):
+        def pack_filenames():
+            dirlist = os.walk(pth, topdown=True)
+            _, _, filenames = next(dirlist)
+            for filename in filenames:
+                yield filename
+        html_top = b'<html><head></head><body><ul>'
+        body = CRLF.join([b"<li>{file}</li>".format(file=file)
+                                    for file in pack_filenames()])
+        html_end = b'</ul></body></html>'
+        message = html_top + body + html_end
+        ctype = b'text/html'
+
+    elif os.path.exists(pth):
+        with open(pth, 'r') as file:
```

Comment:
When opening a file to read, it's best to use `rb` mode (read binary).  If you omit the `b`, then binary data files like images, audio, video and files like pdf may be corrupted by the OS, which may attempt to convert them to human-readable text instead of leaving the bytes of the file intact.